Add k8s inspection capability and dp upgrade e2e tests

darkowlzz opened the pull request:
This change enables the init container to query k8s API and fetch the
necessary information it needs about storageos cluster for performing
any upgrades.

Also, adds e2e tests for dp upgrade in and out of k8s cluster.

darkowlzz committed:
Inspect k8s and fetch the necessary info

This changes enables the init script to query k8s API and fetch the
necessary information it needs for performing any upgrades.

Adds:
- package inspector for inspecting k8s resources.
- package script and runner for scanning for and running scripts.
- mocks for Inspector and Runner interfaces, used in tests.
- env var support. All the scripts are passed all the env vars.

darkowlzz committed:
Add e2e tests for dataplane upgrade.

Adds e2e tests for dataplane upgrade within a k8s cluster and outside of
a k8s cluster.

croomes commented on e2e.sh:
```diff
@@ -35,35 +36,156 @@ run_kind() {
+    sleep 5
+
+    # Patch DaemonSet with new storageos version.
+    kubectl set image ds/storageos-daemonset storageos=storageos/node:1.4.0
```
This is much nicer, I think users will expect to be able to just edit the DS to upgrade the version. :+1:

croomes commented on daemonset.yaml:
```diff
@@ -0,0 +1,149 @@
+  - '*'
+- apiGroups: [""]
+  resources: ["secrets"]
+  verbs: ["create", "get", "list", "delete"]
```
Why does the init container need access to secrets?

darkowlzz replied:
No, this is not for the init container. This cluster role is for the overall need of the DaemonSet. Secret is needed by the node container when it starts.
https://github.com/storageos/cluster-operator/pull/155 has the exact "get" permission for DaemonSet that init container needs.

croomes commented on init.go:
```diff
@@ -1,89 +1,133 @@
+	if storageosImage == "" {
+		log.Println("unknown storageos node image, pass node image with -nodeImage flag.")
+		os.Exit(1)
+	}
```
Maybe we should just use env vars, or make them primary?  Or was there was a reason to also have flags?  Not important...

darkowlzz replied:
No specific reason. Since there's a flag for passing the scripts dir path, I decided to add options for other parameters too. Flags are used in the tests for now. But we can change that if we prefer to use just env vars.

darkowlzz committed:
Reorganize packages.

Add package info containing interface and implementation of info source.
Remove package k8s and move the same code under info/k8s.

domodwyer commented on daemonset.yaml:
```diff
@@ -0,0 +1,155 @@
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
+        - name: DISABLE_TELEMETRY
```
Left in after testing?

I don't think we'd want to start containers with telemetry disabled automatically.

darkowlzz replied:
No, this manifest is only for testing. It's used in e2e test.

domodwyer replied:
Nice! In that case, thanks for disabling it during testing :)

domodwyer commented on script/script.go:
```diff
@@ -0,0 +1,55 @@
+
+	allScripts := []string{}
+
+	err := filepath.Walk(scriptsDir, func(path string, info os.FileInfo, err error) error {
```
Perhaps we can whitelist rather than blacklist? It's a little more robust if we have to add other artefacts (maybe a Go binary / library / etc) that is called by a script.

If so, we could simplify this to a `filepath.Glob("*.sh")` too

darkowlzz replied:
We have executable shell scripts and binaries (dp upgrade binary) here, without any identifiable extension. I wasn't aware that we'll have executable binaries as well. Whitelisting based on file extension won't work because scripts are not just shell scripts in this case. Maybe "scripts" is not a good name, "executables" would have been better?
As far as I know, we'll have single file binary/scripts only for now.

domodwyer replied:
Ah OK, sorry I thought we were replicating the rc init sort of model where we had a collection of shell scripts that were ran in alphabetic order, shelling out as necessary.

No worries.

croomes reviewed (approved):
Looks good, a few comments, none blockers.
